cmd/clusterctl/client/get_kubeconfig_test.go

```diff
@@ -67,3 +75,70 @@ func Test_clusterctlClient_GetKubeconfig(t *testing.T) {
 		})
 	}
 }
+
+const validUserKubeconfig = `
+apiVersion: v1
+kind: Config
+clusters:
+- cluster:
+    certificate-authority-data: randomstring
+    server: https://test-cluster-api.us-east-1.eks.amazonaws.com
+  name: default_mgmt-cluster-control-plane
+contexts:
+- context:
+    cluster: default_mgmt-cluster-control-plane
+    user: default_mgmt-cluster-control-plane-user
+  name: default_mgmt-cluster-control-plane-user@default_mgmt-cluster-control-plane
+current-context: default_mgmt-cluster-control-plane-user@default_mgmt-cluster-control-plane
+preferences: {}
+users
+- name: default_mgmt-cluster-control-plane-user
+  user:
+    client-certificate-data: stuff-cert-data
+    client-key-data: stuff-key-data
+`
+
+func Test_clusterctlClient_GetKubeconfig1(t *testing.T) {
```

Comment:
@ykakarap I have added a temporary test case. It currently fails with the following error:
```sh
=== RUN   Test_clusterctlClient_GetKubeconfig1
    get_kubeconfig_test.go:142: 
        Unexpected error:
            <*errors.withStack | 0xc0002d63c0>: {
                error: <*errors.withMessage | 0xc0005ec080>{
                    cause: <*errors.StatusError | 0xc0003a50e0>{
                        ErrStatus: {
                            TypeMeta: {Kind: "", APIVersion: ""},
                            ListMeta: {
                                SelfLink: "",
                                ResourceVersion: "",
                                Continue: "",
                                RemainingItemCount: nil,
                            },
                            Status: "Failure",
                            Message: "customresourcedefinitions.apiextensions.k8s.io \"clusters.cluster.x-k8s.io\" not found",
                            Reason: "NotFound",
                            Details: {
                                Name: "clusters.cluster.x-k8s.io",
                                Group: "apiextensions.k8s.io",
                                Kind: "customresourcedefinitions",
                                UID: "",
                                Causes: nil,
                                RetryAfterSeconds: 0,
                            },
                            Code: 404,
                        },
                    },
                    msg: "failed to check Cluster API version",
                },
                stack: [0x182dee5, 0x18af66f, 0x18c7316, 0x18c729c, 0x516442, 0x46cf61],
            }
            failed to check Cluster API version: customresourcedefinitions.apiextensions.k8s.io "clusters.cluster.x-k8s.io" not found
        occurred
--- FAIL: Test_clusterctlClient_GetKubeconfig1 (0.00s)
```
While debugging this, I found out that `GetKubeconfig` calls `ProviderInvetory`, which uses the internalClient.proxy instead of the fakeProxy to obtain the objects (in this case the crd). The internal client of the clusterClient that is passed to `GetKubeconfig` during the test does not have any objects in it's proxy, hence the error.
https://github.com/kubernetes-sigs/cluster-api/blob/447bc5be5fb3a437962c1ce18080549612742883/cmd/clusterctl/client/get_kubeconfig.go#L44

The internalClient is created before any objects are added to the fake proxy. The solution that I see is either redefining the internalClient after objects are added, or modifying the `newFakeCluster` to take in an additional argument for proxy.
https://github.com/kubernetes-sigs/cluster-api/blob/447bc5be5fb3a437962c1ce18080549612742883/cmd/clusterctl/client/client_test.go#L208-L231

I cannot say for sure if there is something amiss with any of the helper functions here. Or perhaps I am using the wrong helper functions. What do you think?

Reply:
I think I found out where I was going wrong. I will push the changes soon.